pkg/lifecycle/reconcile.go

```diff
@@ -0,0 +1,185 @@
+}
+
+// ensureCreated - invokes the controller's EnsureCreated method and invokes the statusFunc to update the status of the resource
+func ensureCreated(ctx context.Context, logger logr.Logger, reconciler Reconciler, runtimeObject RuntimeObject, statusFunc Conditioner) (ctrl.Result, error) {
+	logger.Info("ensure created")
+	result, reconcileResult, err := reconciler.EnsureCreated(ctx, runtimeObject)
+	if statusFunc != nil {
+		statusFunc(ctx, runtimeObject)
+	} else {
+		setupCreateState(runtimeObject, reconcileResult, err)
+	}
+	return result, err
+}
+
+// ensureDeleted - invokes the controller's EnsureDeleted method and sets the status of the resource to deleted
+func ensureDeleted(ctx context.Context, logger logr.Logger, reconciler Reconciler, runtimeObject RuntimeObject) (ctrl.Result, error) {
+	logger.Info("ensure deleted")
+	setupDeleteState(runtimeObject, Pending, nil)
+	result, reconcileResult, err := reconciler.EnsureDeleted(ctx, runtimeObject)
+	setupDeleteState(runtimeObject, reconcileResult, err)
+	return result, err
+}
+
+// setupDeleteState - converts the reconcile result to a condition and sets it in the runtimeObject for deletion phase
+func setupDeleteState(runtimeObject RuntimeObject, reconcileResult ReconcileResult, err error) {
+	var condition greenhousev1alpha1.Condition
+	switch reconcileResult {
+	case Success:
+		condition = greenhousev1alpha1.TrueCondition(greenhousev1alpha1.DeleteCondition, DeletedReason, "resource is successfully deleted")
+	case Failed:
+		msg := ""
+		if err != nil {
+			msg = err.Error()
+		}
+		condition = greenhousev1alpha1.FalseCondition(greenhousev1alpha1.DeleteCondition, FailingDeletionReason, "resource deletion failed: "+msg)
+	default:
+		condition = greenhousev1alpha1.FalseCondition(greenhousev1alpha1.DeleteCondition, PendingDeletionReason, "resource deletion is pending")
+	}
+	runtimeObject.SetCondition(condition)
+}
+
+// setupCreateState - if statusFunc is not passed to reconciler then the default status conditions are set in runtimeObject
+func setupCreateState(runtimeObject RuntimeObject, reconcileResult ReconcileResult, err error) {
+	var condition greenhousev1alpha1.Condition
+	switch reconcileResult {
+	case Success:
+		condition = greenhousev1alpha1.TrueCondition(greenhousev1alpha1.ReadyCondition, CreatedReason, "resource is successfully created")
+	case Failed:
+		msg := ""
+		if err != nil {
+			msg = err.Error()
+		}
+		condition = greenhousev1alpha1.FalseCondition(greenhousev1alpha1.ReadyCondition, FailingCreationReason, "resource creation failed"+msg)
+	default:
+		condition = greenhousev1alpha1.UnknownCondition(greenhousev1alpha1.ReadyCondition, PendingCreationReason, "resource creation is pending")
+	}
+	runtimeObject.SetCondition(condition)
+}
+
+// patchStatus - patches the status of the resource with the new status and returns the reconcile error
+// TODO: implement event recorder to fire ready condition change events
```

Comment:
Great work Abhi! 🚀  I went thru your PR to keep myself updated
Have only one question.
Is this TODO still relevant for this PR?

Reply:
Hi @ibakshay No not for this PR, there will be a follow up task, can you give me a  `suggestion` to remove the comment?